internal/services/autorecap/autorecap.go

```diff
@@ -403,14 +403,61 @@ func (m *AutoRecapService) summarize(chatID int64, options *ent.TelegramChatReca
 				msg.ReplyMarkup = inlineKeyboardMarkup
 			}
 
-			_, err = m.botService.Send(msg)
+			sent_msg, err := m.botService.Send(msg)
 			if err != nil {
 				m.logger.Error("failed to send chat histories recap",
 					zap.Int64("chat_id", chatID),
 					zap.Int("auto_recap_rates", options.AutoRecapRatesPerDay),
 					zap.Error(err),
 				)
 			}
+
+			// Check whether the first message of the batch needs to be pinned
+			if i != 0 || !options.PinAutoRecapMessage {
+				err = m.chathistories.SaveOneTelegramSentMessage(&sent_msg, false)
+				if err != nil {
+					m.logger.Error("failed to save one telegram sent message",
+						zap.Int64("chat_id", chatID),
+						zap.Error(err))
+				}
+
+				return
+			}
+
+			// Unpin the last pinned message
+			lastPinnedMessage, err := m.chathistories.FindLastTelegramPinnedMessage(chatID)
+			if err != nil {
+				m.logger.Error("failed to find last pinned message",
+					zap.Int64("chat_id", chatID),
+					zap.Error(err),
+				)
+			}
+
+			if err = m.botService.UnpinChatMessage(tgbot.NewUnpinChatMessageConfig(chatID, lastPinnedMessage.MessageID)); err != nil {
+				m.logger.Error("failed to unpin chat message",
+					zap.Int64("chat_id", chatID),
+					zap.Error(err),
+				)
+			}
+
+			if err = m.chathistories.UpdatePinnedMessage(lastPinnedMessage.ChatID, lastPinnedMessage.MessageID, false); err != nil {
+				m.logger.Error("failed to save one telegram sent message",
+					zap.Int64("chat_id", chatID),
+					zap.Error(err))
+			}
+
+			if err = m.botService.PinChatMessage(tgbot.NewPinChatMessageConfig(chatID, sent_msg.MessageID)); err != nil {
+				m.logger.Error("failed to pin chat message",
+					zap.Int64("chat_id", chatID),
+					zap.Error(err),
+				)
+			}
+
+			if err = m.chathistories.SaveOneTelegramSentMessage(&sent_msg, true); err != nil {
+				m.logger.Error("failed to save one telegram sent message",
+					zap.Int64("chat_id", chatID),
+					zap.Error(err))
+			}
```

Comment:
You can use the similar technic like what you have written with https://github.com/nekomeowww/insights-bot/pull/206/files#diff-99fbab3f020962dc5770c6e55c402059116a0a019d7c9a243de8db596ed8b6afR256-R260 as well as https://github.com/nekomeowww/insights-bot/blob/0ec307bb387b3cf3f28fc623d36d57fa7376667c/internal/services/autorecap/autorecap.go#L91-L120 with  https://github.com/nekomeowww/fo?tab=readme-ov-file#may0-6 to better handle the errors like this:

```suggestion
			may := fo.NewMay0().Use(func(err error, messageArgs ...any) {
				if len(messageArgs) == 0 {
					m.logger.Error(err.Error())
					return
				}

				prefix, _ := messageArgs[0].(string)
				if len(messageArgs) == 1 {
					m.logger.Error(prefix, zap.Error(err))
					return
				}

				fields := make([]zap.Field, 0)
				fields = append(fields, zap.Error(err))

				for i, v := range messageArgs[1:] {
					field, ok := v.(zap.Field)
					if !ok {
						fields = append(fields, zap.Any(fmt.Sprintf("error_field_%d", i), field))
					} else {
						fields = append(fields, field)
					}
				}

				m.logger.Error(prefix, fields...)
			})

			// Unpin the last pinned message
			lastPinnedMessage, err := m.chathistories.FindLastTelegramPinnedMessage(chatID)
			if err != nil {
				m.logger.Error("failed to find last pinned message",
					zap.Int64("chat_id", chatID),
					zap.Error(err),
				)
			}

			may.Invoke(m.botService.UnpinChatMessage(tgbot.NewUnpinChatMessageConfig(chatID, lastPinnedMessage.MessageID)), "failed to unpin chat message", zap.Int64("chat_id", chatID), zap.Int("message_id", lastPinnedMessage.MessageID))
			may.Invoke(m.chathistories.UpdatePinnedMessage(lastPinnedMessage.ChatID, lastPinnedMessage.MessageID, false), "failed to save one telegram sent message", zap.Int64("chat_id", lastPinnedMessage.ChatID), zap.Int("message_id", lastPinnedMessage.MessageID))
			may.Invoke(m.botService.PinChatMessage(tgbot.NewPinChatMessageConfig(chatID, sentMsg.MessageID)), "failed to pin chat message", zap.Int64("chat_id", chatID), zap.Int("message_id", sentMsg.MessageID))
			may.Invoke(m.chathistories.SaveOneTelegramSentMessage(&sentMsg, true), "failed to save one telegram sent message")
```

In fact I think a universal and general implementation for middleware of `fo.NewMay0()` can be extracted later and contribute to [github.com/nekomeowww/fo](github.com/nekomeowww/fo) since I think this might be a pretty straightforward way to handle the optional errors with logs. What do you think?

Reply:
I think this idea is cool. However, I'm new to Golang and have little knowledge about the best practice to handle errors. So it is up to you.
Besides, https://github.com/nekomeowww/insights-bot/pull/206/files#diff-45a36e95ac6c3baa972b255d242ba46f440324d91d294cf71318fd5d6e2f6172R416
when unpinning a message, it is required to check whether it is the first message in this batch. I will submit a new commit later following the lo style.

Reply:
> I'm new to Golang and have little knowledge about the best practice to handle errors.

Relax, it's ok. The `fo.NewMay` is just another little utilities that I invented by myself to handle these log related errors. This is not something that so practical to learn and enforce with.

Good job as a beginner! Congratulations for your amazing contribution!